jax/_src/random.py

```diff
@@ -1491,6 +1491,105 @@ def _t(key, df, shape, dtype) -> Array:
+      shape. Must be broadcast-compatible with ``df``. The default (None)
+      produces a result shape equal to ``df.shape``.
+    dtype: optional, a float dtype for the returned values (default float64 if
+      jax_enable_x64 is true, otherwise float32).
+
+  Returns:
+    A random array with the specified dtype and with shape given by ``shape`` if
+    ``shape`` is not None, or else by ``df.shape``.
+  """
+  key, _ = _check_prng_key(key)
+  if not dtypes.issubdtype(dtype, np.floating):
+    raise ValueError("dtype argument to `chisquare` must be a float "
+                     f"dtype, got {dtype}")
+  dtype = dtypes.canonicalize_dtype(dtype)
+  if shape is not None:
+    shape = core.canonicalize_shape(shape)
+  return _chisquare(key, df, shape, dtype)
+
+@partial(jit, static_argnums=(2, 3), inline=True)
+def _chisquare(key, df, shape, dtype) -> Array:
+  if shape is None:
+    shape = np.shape(df)
+  else:
+    _check_shape("chisquare", shape, np.shape(df))
+  df = lax.convert_element_type(df, dtype)
+  two = _lax_const(df, 2)
+  half_df = lax.div(df, two)
+  log_g = loggamma(key, a=half_df, shape=shape, dtype=dtype)
+  chi2 = lax.mul(jnp.exp(log_g), two)
+  return chi2
+
+
+def f(key: KeyArray,
+      dfnum: RealArray,
+      dfden: RealArray,
+      shape: Optional[Shape] = None,
+      dtype: DTypeLikeFloat = dtypes.float_) -> Array:
+  """Sample F-distribution random values with given shape and float dtype.
+
+  Args:
+    key: a PRNG key used as the random key.
+    dfnum: a float or array of floats broadcast-compatible with ``shape``
+      representing the numerator's ``df`` of the distribution.
+    dfden: a float or array of floats broadcast-compatible with ``shape``
+      representing the denominator's ``df`` of the distribution.
+    shape: optional, a tuple of nonnegative integers specifying the result
+      shape. Must be broadcast-compatible with ``dfnum`` and ``dfden``.
+      The default (None) produces a result shape equal to ``dfnum.shape``,
+      and ``dfden.shape``.
+    dtype: optional, a float dtype for the returned values (default float64 if
+      jax_enable_x64 is true, otherwise float32).
+
+  Returns:
+    A random array with the specified dtype and with shape given by ``shape`` if
+    ``shape`` is not None, or else by ``df.shape``.
+  """
+  key, _ = _check_prng_key(key)
+  if not dtypes.issubdtype(dtype, np.floating):
+    raise ValueError("dtype argument to `f` must be a float "
+                     f"dtype, got {dtype}")
+  dtype = dtypes.canonicalize_dtype(dtype)
+  if shape is not None:
+    shape = core.canonicalize_shape(shape)
+  return _f(key, dfnum, dfden, shape, dtype)
+
+@partial(jit, static_argnums=(3, 4), inline=True)
+def _f(key, dfnum, dfden, shape, dtype) -> Array:
+  # broadcast dfnum and dfden to generate chi2 rvs
+  dfden, dfnum = jnp.broadcast_arrays(dfden, dfnum)
+  if shape is None:
+    shape = dfden.shape
+  else:
+    shape = dfden.shape + shape
```

Comment:
This seems like surprising shape semantics. Consider the analogous numpy API:
```python
import numpy as np
result = np.random.f(np.ones(5), 2, size=(5,))
print(result.shape)
# (5,)
```
Here, doing the equivalent would lead to the output having shape `(5, 5)`.

We should probably handle the `shape` parameter similarly to other functions in this file, e.g. https://github.com/google/jax/blob/f66f6ec98a2bceeb7f609cbfdbeee81dc74ad32f/jax/_src/random.py#L644-L647

Reply:
Yeah, that API seems odd to me as well.

Reply:
In general, it looks like the random submodule is very inconsistent in its handling of shapes and dtypes, and for the most parts the tests don't cover this. We should probably just overhaul it all in a single pass.

Reply:
Ok i think i'll raise an issue and open another pr for double_side_maxwellafter modifying this pr. ：）

Reply:
I think rather than a PR just for `double_side_maxwell`, we need a revamp of most of the `jax.random` module to make the API consistent and robust. That gets tricky because of backward compatibility and API deprecation issues; I'll plan to look into that soon.